src/tokens/ERC1155/ERC1155Sale.sol

```diff
@@ -0,0 +1,244 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.17;
+
+import {IERC1155Sale} from "./IERC1155Sale.sol";
+import {ERC1155Supply, ERC1155Token} from "./ERC1155Supply.sol";
+import {ERC1155SaleErrors} from "./ERC1155SaleErrors.sol";
+import {WithdrawControlled, AccessControl, SafeERC20, IERC20} from "../common/WithdrawControlled.sol";
+
+contract ERC1155Sale is
+    IERC1155Sale,
+    ERC1155SaleErrors,
+    ERC1155Supply,
+    WithdrawControlled
+{
+    bytes32 public constant MINT_ADMIN_ROLE = keccak256("MINT_ADMIN_ROLE");
+
+    bytes4 private constant _ERC20_TRANSFERFROM_SELECTOR =
+        bytes4(keccak256(bytes("transferFrom(address,address,uint256)")));
+
+    bool private _initialized;
+
+    // ERC20 token address for payment. address(0) indicated payment in ETH.
+    address private _paymentToken;
+
+    SaleDetails private _globalSaleDetails;
+    mapping(uint256 => SaleDetails) private _tokenSaleDetails;
+
+    /**
+     * Initialize the contract.
+     * @param owner Owner address.
+     * @param tokenName Token name.
+     * @param tokenBaseURI Base URI for token metadata.
+     * @dev This should be called immediately after deployment.
+     */
+    function initialize(address owner, string memory tokenName, string memory tokenBaseURI) public virtual override {
+        if (_initialized) {
+            revert InvalidInitialization();
+        }
+        ERC1155Token.initialize(owner, tokenName, tokenBaseURI);
+
+        name = tokenName;
+        baseURI = tokenBaseURI;
+        _setupRole(DEFAULT_ADMIN_ROLE, owner);
+        _setupRole(MINT_ADMIN_ROLE, owner);
+        _setupRole(WITHDRAW_ROLE, owner);
+
+        _initialized = true;
+    }
+
+    /**
+     * Checks if the current block.timestamp is out of the give timestamp range.
+     * @param _startTime Earliest acceptable timestamp (inclusive).
+     * @param _endTime Latest acceptable timestamp (exclusive).
+     * @dev A zero endTime value is always considered out of bounds.
+     */
+    function blockTimeOutOfBounds(uint256 _startTime, uint256 _endTime) private view returns (bool) {
+        // 0 end time indicates inactive sale.
+        return _endTime == 0 || block.timestamp < _startTime || block.timestamp >= _endTime; // solhint-disable-line not-rely-on-time
+    }
+
+    /**
+     * Checks the sale is active and takes payment.
+     * @param _tokenIds Token IDs to mint.
+     * @param _amounts Amounts of tokens to mint.
+     */
+    function _payForActiveMint(uint256[] memory _tokenIds, uint256[] memory _amounts) private {
```

Comment:
We probably need the ability to add a mint limit per token ID. Should be optional, i.e. if limit is 0, then there is no limit, otherwise it's the limit specified.

Reply:
Is that per transaction / wallet? There is a supply cap limit per token already